### PR TITLE
User Notification Preferences

### PR DESCRIPTION
**Pull request by LomyW**
some changes to save the `notification preferences` and mutate them.

### CODE REVIEW

File: common/entity/Person.ts [hunk 1]
```diff
@@ -66,6 +66,16 @@ export abstract class Person {
...
+
+    @Column({
+        type: 'timestamp',
+        nullable: true,
+        default: new Date(),
+    })
+    lastTimeCheckedNotifications: Date;
+
+    @Column({ nullable: true, type: 'json' })
+    notificationPreferences: any;
```
Comment (jeffersondarcy) [inline common/entity/Person.ts:78]: why not string?

File: common/migration/1669676905990-AddLastCheckedNotifications.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import {MigrationInterface, QueryRunner} from "typeorm";
+
+export class AddLastCheckedNotifications1669676905990 implements MigrationInterface {
+    name = 'AddLastCheckedNotifications1669676905990'
+
+    public async up(queryRunner: QueryRunner): Promise<void> {
+        await queryRunner.query(`ALTER TABLE "pupil" ADD "lastTimeCheckedNotifications" TIMESTAMP DEFAULT '"2022-11-28T23:08:26.067Z"'`);
```
Comment (jeffersondarcy) [inline common/migration/1669676905990-AddLastCheckedNotifications.ts:7]: why not default NULL?

File: common/entity/Person.ts [hunk 3]
```diff
@@ -66,6 +66,16 @@ export abstract class Person {
         default: false,
     })
     isRedacted: boolean;
+
+    @Column({
+        type: 'timestamp',
+        nullable: true,
+        default: new Date(),
```
Comment (jeffersondarcy) [inline common/entity/Person.ts:73]: I am not sure about default. When the column is created everything should be NULL, but on writing current timestamp is right

File: common/user/session.ts [hunk 4]
```diff
@@ -15,6 +15,8 @@ export const UNAUTHENTICATED_USER = {
     firstname: '',
     lastname: '',
     userID: '-/-',
+    lastTimeCheckedNotifications: new Date(),
```
Comment (jeffersondarcy) [inline common/user/session.ts:18]: I think it's wrong - all present notifications would be older than the current timestamp

File: common/migration/1669676905990-AddLastCheckedNotifications.ts (hunk 2 above)
Comment (jeffersondarcy) [inline common/migration/1669676905990-AddLastCheckedNotifications.ts:7]: I think you need to alter every table in two steps, first using TIMESTAMP DEFAULT NULL to fill the column with null values, than alter DEFAULT to current timestamp for future mutating queries https://www.postgresqltutorial.com/postgresql-date-functions/postgresql-current_timestamp/

File: common/util/preferences-utils.ts [hunk 5]
```diff
@@ -0,0 +1,9 @@
+type ChannelPreference = {
```
Comment (jeffersondarcy) [inline common/util/preferences-utils.ts:1]: I would call it `Channels`

File: common/util/preferences-utils.ts [hunk 6]
```diff
@@ -0,0 +1,9 @@
...
+    email: boolean;
+};
+
+type PreferenceCategories = 'chat' | 'match' | 'course' | 'appointment' | 'survey' | 'news';
```
Comment (jeffersondarcy) [inline common/util/preferences-utils.ts:5]: I would call it `MessageTypes` or `NotificationTypes` or maybe `MessageEvents`

**Review by jeffersondarcy — CHANGES_REQUESTED** [review]
please check my comments on timestamp defaults

File: common/entity/Person.ts (hunk 1 above)
Comment (LomyW) [reply common/entity/Person.ts:78]: Idk, I think this code might be from @Jonasdoubleyou. Maybe the type `NotificationPrefereces | null` could be used here, but I'm not sure.

File: common/migration/1669676905990-AddLastCheckedNotifications.ts (hunk 2 above)
Comment (LomyW) [reply common/migration/1669676905990-AddLastCheckedNotifications.ts:7]: This was automatically generated, I wouldn't touch this.

File: common/user/session.ts (hunk 4 above)
Comment (LomyW) [reply common/user/session.ts:18]: This is an unauthorized user, I just set a value to satisfy TS an unauthorized user shouldn't have any notifications.

File: common/entity/Person.ts (hunk 1 above)
Comment (Jonasdoubleyou) [reply common/entity/Person.ts:78]: Ah yeah, a proper type could also be used here, was too lazy 😅

File: common/entity/Person.ts (hunk 3 above)
Comment (Jonasdoubleyou) [reply common/entity/Person.ts:73]: Ah, maybe we should set it to the timestamp that was generated in the migration, otherwise TypeORM might generate a new migration each time ... (haven't checked this though)

File: common/entity/Person.ts [hunk 7]
```diff
@@ -66,6 +67,16 @@ export abstract class Person {
...
+
+    @Column({
+        type: 'timestamp',
+        nullable: true,
+        default: new Date(0),
+    })
+    lastTimeCheckedNotifications: Date;
+
+    @Column({ nullable: true, type: 'json' })
+    notificationPreferences: Prisma.JsonValue;
```
Comment (jeffersondarcy) [inline common/entity/Person.ts:79]: @Jonasdoubleyou is Prisma.JsonValue the correct type?

Comment (Jonasdoubleyou) [reply common/entity/Person.ts:79]: No, NotificationPreferences was better.

Comment (Jonasdoubleyou) [reply common/entity/Person.ts:79]: The entity/* Files are from TypeORM, maybe better not to mix in Prisma Types here

File: common/notification/index.ts [hunk 8]
```diff
@@ -261,16 +262,39 @@ async function deliverNotification(
...
+                if (enabledChannels[channel] === true) {
+                    activeChannelForNotificationType.push(channel);
+                }
+            }
+        }
+
+        const channelsToSendTo = channels
+            .filter((it) => it.canSend(notification, user))
+            .filter((c) => {
+                const compareValue = c.type === 'mailjet' ? 'email' : c.type;
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:291]: Uh, parity please between preferences and the channel names. Feel free to rename the channel (I guess "email" is a better name anyways)

File: common/notification/index.ts [hunk 9]
```diff
@@ -261,16 +262,39 @@ async function deliverNotification(
...
+
         // TODO: Check if user silenced this notification
+        const { notificationPreferences } = await getStudentOrPupil(user);
+
+        if (notificationPreferences && typeof notificationPreferences === 'string') {
+            const enabledChannels: { [channel: string]: boolean } = JSON.parse(notificationPreferences)[messageType];
+
+            for (const channel in enabledChannels) {
+                if (enabledChannels[channel] === true) {
+                    activeChannelForNotificationType.push(channel);
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:283]: I don't see the point of `activeChannelForNotificationType`, what benefit does this array have over the `enabledChannels` dictionary?

File: common/notification/index.ts [hunk 10]
```diff
@@ -261,16 +262,39 @@ async function deliverNotification(
...
+                }
+            }
+        }
+
+        const channelsToSendTo = channels
+            .filter((it) => it.canSend(notification, user))
+            .filter((c) => {
+                const compareValue = c.type === 'mailjet' ? 'email' : c.type;
+                return activeChannelForNotificationType.includes(compareValue);
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:292]: `enabledChannels[ channel.type ]` ...

File: common/notification/index.ts [hunk 11]
```diff
@@ -261,16 +262,39 @@ async function deliverNotification(
...
+        const { messageType } = getMessage(concreteNotification);
+
+        // default channel is inapp
+        const activeChannelForNotificationType: string[] = ['inapp'];
+
         // TODO: Check if user silenced this notification
+        const { notificationPreferences } = await getStudentOrPupil(user);
+
+        if (notificationPreferences && typeof notificationPreferences === 'string') {
+            const enabledChannels: { [channel: string]: boolean } = JSON.parse(notificationPreferences)[messageType];
```
[the remark this answers is not in the document]
Comment (Jonasdoubleyou) [reply common/notification/index.ts:279]: Also this should gracefully fall back when `messageType` is not present in the preferences

File: common/notification/index.ts [hunk 12]
```diff
@@ -261,16 +262,39 @@ async function deliverNotification(
...
+            for (const channel in enabledChannels) {
+                if (enabledChannels[channel] === true) {
+                    activeChannelForNotificationType.push(channel);
+                }
+            }
+        }
+
+        const channelsToSendTo = channels
+            .filter((it) => it.canSend(notification, user))
+            .filter((c) => {
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:290]: I'm not really a fan of one character variable names, how about `channel`?

File: common/notification/index.ts [hunk 13]
```diff
@@ -261,16 +262,39 @@ async function deliverNotification(
...
             await triggerHook(notification.hookID, user);
         }
 
+        const { messageType } = getMessage(concreteNotification);
+
+        // default channel is inapp
+        const activeChannelForNotificationType: string[] = ['inapp'];
+
         // TODO: Check if user silenced this notification
+        const { notificationPreferences } = await getStudentOrPupil(user);
```
[the remark this answers is not in the document]
Comment (Jonasdoubleyou) [reply common/notification/index.ts:276]: Here `queryUser({ notificationPreferences: true })` would also work

File: common/notification/index.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jonasdoubleyou) [reply common/notification/index.ts:279]: Also I'd prefer typing whatever `JSON.parse` returns before then accessing it's members.

File: graphql/authorizations.ts [hunk 14]
```diff
@@ -240,7 +240,8 @@ export const authorizationModelEnhanceMap: ModelsEnhanceMap = {
             state: adminOrOwner,
             teacherEmailAddress: adminOrOwner,
             coduToken: adminOrOwner,
-
+            lastTimeCheckedNotifications: adminOrOwner,
```
Comment (timidak) [inline graphql/authorizations.ts:243]: Why is this here as well? 
Something off with the history?

**Review by timidak — COMMENTED** [review]
Just a few questions 😄

File: common/notification/index.ts [hunk 15]
```diff
@@ -261,21 +282,26 @@ async function deliverNotification(
...
 
         if (notification.hookID) {
             await triggerHook(notification.hookID, user);
         }
 
-        // TODO: Check if user silenced this notification
+        // default channel is webApp is always enabled
+        const enabledChannels: Array<Channel> = [channels['webApp']];
+
+        const messageTypeChannelPreferences = await getNotificationChannelPreferences(user, concreteNotification);
```
Comment (timidak) [inline common/notification/index.ts:293]: channelPreferencesForMessageType?

File: graphql/authorizations.ts (hunk 14 above)
Comment (jeffersondarcy) [reply graphql/authorizations.ts:243]: one is for Pupil and another for Student :)

File: common/notification/index.ts [hunk 16]
```diff
@@ -243,6 +246,21 @@ async function createConcreteNotification(
     return concreteNotification;
 }
 
+const getNotificationChannelPreferences = async (user: User, concreteNotification: ConcreteNotification): Promise<Channels> => {
+    const { messageType } = getMessage(concreteNotification);
```
Comment (Jonasdoubleyou) [inline common/notification/index.ts:250]: So the message type now determines the notification channel? I don't like that. This should be notification.type.

Comment (jeffersondarcy) [reply common/notification/index.ts:250]: currently we only map types in templates. as we are moving templates to db anyway we could add the type field to notification table. would that be a good solution

Comment (Jonasdoubleyou) [reply common/notification/index.ts:250]: Yes, can also be done as part of LF-60 if you want to bring this PR in

**Review by Jonasdoubleyou — APPROVED** [review]

File: common/notification/index.ts (hunk 16 above)
Comment (jeffersondarcy) [reply common/notification/index.ts:250]: so I will keep this solution till change it in LF-60, okay?
